File: MoonEngine/Base/src/main/java/com/greentree/engine/moon/base/sorter/MethodSorter.java
package com.greentree.engine.moon.base.sorter;

import com.greentree.commons.graph.DirectedGraph;
import com.greentree.commons.graph.Graph;
import com.greentree.commons.util.iterator.IteratorUtil;

import java.util.List;

public interface MethodSorter {

    default <T> void sort(List<T> modules, String method) {
        @SuppressWarnings("unchecked")
        var graph = (DirectedGraph<T>) buildDependencyGraph(modules, method);
        final var cycle = graph.getCycleFinder().getCycles();
        if (!IteratorUtil.isEmpty(cycle))
            throw new IllegalArgumentException(modules + " " + method + " has cycle " + cycle);
        for (var v : modules)
            graph.add(v);
        modules.clear();
        graph.getTopologicalSort(modules);
    }

    <T> Graph<? extends T> buildDependencyGraph(List<? extends T> modules, String method);

}

File: MoonEngine/Base/src/main/java/com/greentree/engine/moon/base/sorter/MethodSorter.kt
package com.greentree.engine.moon.base.sorter

import com.greentree.commons.graph.DirectedGraph
import com.greentree.commons.graph.Graph

interface MethodSorter {

	fun <T> sort(modules: MutableList<T>, method: String) {
		val graph = buildDependencyGraph<T>(modules, method) as DirectedGraph<T>
		val cycle = graph.cycleFinder.cycles
		require(cycle.isEmpty()) { "$method has cycle ${cycle.first()}" }
		for(v in modules) graph.add(v)
		modules.clear()
		graph.getTopologicalSort(modules)
	}

	fun <T> buildDependencyGraph(modules: List<T>, method: String): Graph<out T>
}
